app/src/org/commcare/views/widgets/CalendarFragment.java

```diff
@@ -0,0 +1,342 @@
+package org.commcare.views.widgets;
+
+import android.app.Activity;
+import android.content.DialogInterface;
+import android.content.Context;
+import android.content.pm.ActivityInfo;
+import android.content.res.Configuration;
+import android.graphics.Rect;
+import android.os.Bundle;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.view.Window;
+import android.widget.AdapterView;
+import android.widget.ArrayAdapter;
+import android.widget.Button;
+import android.widget.GridView;
+import android.widget.ImageButton;
+import android.widget.LinearLayout;
+import android.widget.Spinner;
+import android.widget.TextView;
+import org.javarosa.core.model.data.DateData;
+import org.commcare.dalvik.R;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Created by Saumya on 5/29/2016.
+ * DialogFragment for a popup calendar icon
+ * Uses support library for compatibility with pre-honeycomb devices
+ * Layout and logic inspired by Ahmed Al-Amir at https://www.toptal.com/android/android-customization-how-to-build-a-ui-component-that-does-what-you-want
+ */
+public class CalendarFragment extends android.support.v4.app.DialogFragment {
+
+    protected GridView myGrid;
+    protected ImageButton cancel;
+    protected Button today;
+    private Spinner monthSpinner;
+    private Spinner yearSpinner;
+    protected LinearLayout myLayout;
+
+    protected Calendar calendar;
+    protected CalendarCloseListener myListener;
+
+    protected long todaysDateInMillis;
+    protected static final int DAYSINWEEK = 7;
+    protected static final String TIME = "TIME";
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState){
+        getDialog().getWindow().requestFeature(Window.FEATURE_NO_TITLE);
+        inflateView(inflater, container);
+
+        if(savedInstanceState != null && savedInstanceState.containsKey(TIME)){
+            calendar = Calendar.getInstance();
+            calendar.setTimeInMillis(savedInstanceState.getLong(TIME));
+        }
+
+        disableScreenRotation();
+        initDisplay();
+        initWeekDays();
+        initOnClick();
+        refresh();
+        setWindowSize();
+
+        return myLayout;
+    }

```

Comment:
I try to order my class methods in terms of usage. I've found that to be a pattern that leads to pretty good readability. 
So for instance, `onCreateView` is one of the most important method, so it comes first in the class. This method calls `N` other methods, which should be defined right below, so that when I read `onCreateView`, I can look right below it to follow what the helper methods are doing. If one of those helper methods makes use of another helper method, I usually put it right below (so it is like a depth first traversal).